File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> mp;
        int ans = 0;
        int start = 0;

        for(int i = 0; i < s.length(); i++) {
            if(mp.find(s[i]) != mp.end() && mp[s[i]] >= start) {
                start = mp[s[i]] + 1;
            }
            mp[s[i]] = i;
            ans = max(ans, i - start + 1);
        }

        return ans;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int start = 0, end = 0;
        HashSet<Character> st = new HashSet<>();
        int res = 0;
        int curr = 0;
        while(end < s.length()){
            if(st.contains(s.charAt(end))){
                curr = end - start;
                res = Math.max(res, curr);
                start++;
                end = start;
                st.clear();
            }
            st.add(s.charAt(end));
            end++;
        }
        curr = end - start;
        res = Math.max(res, curr);
        return res;
    }
}